tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/BuilderUtil.scala

```diff
@@ -91,18 +93,23 @@ object BuilderUtil {
       case (cmp, (variable, set)) =>
         for {
           seq <- cmp
+          usedBefore <- allUsed
           setEx <- set
-          usedInSet <- allUsed // variable_i may not appear as bound or free in set_i
+          usedInSetOrBefore <- allUsed // variable_i may not appear as bound or free in set_i
+          usedInSet = usedInSetOrBefore -- usedBefore
           varEx <- variable
-          _ = require(varEx.isInstanceOf[NameEx])
-          _ <- markAsBound(varEx)
-          // variable_i is shadowed iff boundAfterVar \subseteq usedInSet \union boundAfterBodyEx
-          boundAfterVar <- allBound
-        } yield
-          if (boundAfterVar.subsetOf(usedInSet.union(boundAfterBodyEx))) {
+          // we delay marking as bound, to not interfere with other variable-set pairs
+        } yield {
+          require(varEx.isInstanceOf[NameEx], s"varEx = $varEx must be a variable name.")
+          // variable_i is shadowed iff boundVar \in usedInSet \union boundAfterBodyEx
+          val boundVar = varEx.asInstanceOf[NameEx].name
+          if (usedInSet.union(boundAfterBodyEx).contains(boundVar))
```

Comment:
These changes don't seem like they can actually be related to the stated purpose of this PR. iiuc, the PR says it is about making a pretty simple design change, that should mostly be mechanical, but these changes look like pretty subtle adjustments to the business logic of key operations. 

If I am reading this right, could you please refactor this PR into to separate parts? One that makes the design/api change about how the unsafe builders are accessed and one that makes the adjustments to the business logic and errors etc? 

It's a quite large PR, and I think factoring it out into those different parts would make review much easier and more useful.

Reply:
Ah yeah. I had totally missed that comment. My bad.

Ideally these would be separate PRs imo, assuming the overhead is not too great. Otherwise, I agree that updating the pr title description will help clarify for reviewers.